File: modules/expo-audio-engine/ios/ExpoAudioEngineModule.swift
/**
 * iOS Native Audio Engine Module
 * 
 * Uses AVAudioEngine for low-latency audio synthesis.
 * Implements real-time tone generation, sweeps, and multi-tone playback.
 * 
 * IMPLEMENTATION GUIDE:
 * 
 * 1. Setup AVAudioEngine with AVAudioPlayerNode
 * 2. Generate PCM buffers for different waveforms
 * 3. Use AVAudioSourceNode for real-time synthesis
 * 4. Apply AVAudioUnitEQ for frequency shaping if needed
 * 5. Monitor audio session interruptions
 * 
 * Key Classes:
 * - AVAudioEngine: Main audio graph
 * - AVAudioPlayerNode: Playback node
 * - AVAudioSourceNode: Real-time synthesis (iOS 13+)
 * - AVAudioPCMBuffer: Audio buffer
 * - AVAudioSession: Audio session management
 * 
 * Performance Tips:
 * - Use preferred sample rate (48kHz typical)
 * - Buffer size: 512 or 1024 samples for low latency
 * - Pre-allocate buffers to avoid runtime allocation
 * - Use vDSP framework for efficient DSP operations
 */

import AVFoundation
import ExpoModulesCore

public class ExpoAudioEngineModule: Module {
  private var audioEngine: AVAudioEngine?
  private var playerNode: AVAudioPlayerNode?
  private var isPlaying = false
  private var sampleRate: Double = 48000
  private var bufferSize: AVAudioFrameCount = 512
  
  // MARK: - Module Definition
  
  public func definition() -> ModuleDefinition {
    Name("ExpoAudioEngineModule")
    
    // Initialize audio engine
    AsyncFunction("initialize") { (config: [String: Any], promise: Promise) in
      do {
        try self.initializeAudioEngine(config: config)
        promise.resolve(nil)
      } catch {
        promise.reject("INIT_ERROR", error.localizedDescription)
      }
    }
    
    // Play tone
    AsyncFunction("playTone") { (params: [String: Any], promise: Promise) in
      do {
        try self.playTone(params: params)
        promise.resolve(nil)
      } catch {
        promise.reject("PLAY_ERROR", error.localizedDescription)
      }
    }
    
    // Play sweep
    AsyncFunction("playSweep") { (params: [String: Any], promise: Promise) in
      do {
        try self.playSweep(params: params)
        promise.resolve(nil)
      } catch {
        promise.reject("SWEEP_ERROR", error.localizedDescription)
      }
    }
    
    // Play multi-tone
    AsyncFunction("playMultiTone") { (params: [String: Any], promise: Promise) in
      do {
        try self.playMultiTone(params: params)
        promise.resolve(nil)
      } catch {
        promise.reject("MULTITONE_ERROR", error.localizedDescription)
      }
    }
    
    // Stop playback
    AsyncFunction("stop") { (promise: Promise) in
      self.stopPlayback()
      promise.resolve(nil)
    }
    
    // Dispose
    AsyncFunction("dispose") { (promise: Promise) in
      self.disposeAudioEngine()
      promise.resolve(nil)
    }
  }
  
  // MARK: - Audio Engine Setup
  
  private func initializeAudioEngine(config: [String: Any]) throws {
    // Configure audio session
    let audioSession = AVAudioSession.sharedInstance()
    try audioSession.setCategory(.playback, mode: .default, options: [])
    try audioSession.setActive(true)
    
    // Extract config
    if let rate = config["sampleRate"] as? Double {
      sampleRate = rate
    }
    if let buffer = config["bufferSize"] as? Int {
      bufferSize = AVAudioFrameCount(buffer)
    }
    
    // Create audio engine
    audioEngine = AVAudioEngine()
    playerNode = AVAudioPlayerNode()
    
    guard let engine = audioEngine, let player = playerNode else {
      throw NSError(domain: "AudioEngine", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to create audio engine"])
    }
    
    // Attach player node
    engine.attach(player)
    
    // Configure audio format
    let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!
    
    // Connect nodes
    engine.connect(player, to: engine.mainMixerNode, format: format)
    
    // Start engine
    try engine.start()
  }
  
  // MARK: - Playback Methods (Stubs)
  
  private func playTone(params: [String: Any]) throws {
    // TODO: Implement tone generation
    // 1. Extract parameters (frequency, duration, gain, waveform)
    // 2. Generate PCM buffer with waveform
    // 3. Apply envelope (fadeIn/fadeOut)
    // 4. Schedule buffer on playerNode
    // 5. Start playback
  }
  
  private func playSweep(params: [String: Any]) throws {
    // TODO: Implement frequency sweep
    // 1. Extract parameters (fromHz, toHz, duration, logarithmic)
    // 2. Generate swept waveform
    // 3. Use AVAudioSourceNode for real-time generation
    // 4. Apply gain envelope
    // 5. Play for specified duration
  }
  
  private func playMultiTone(params: [String: Any]) throws {
    // TODO: Implement multi-tone synthesis
    // 1. Extract frequency array and gains
    // 2. Generate each tone separately
    // 3. Mix buffers (sum samples)
    // 4. Normalize to prevent clipping
    // 5. Play mixed buffer
  }
  
  private func stopPlayback() {
    playerNode?.stop()
    isPlaying = false
  }
  
  private func disposeAudioEngine() {
    stopPlayback()
    audioEngine?.stop()
    audioEngine = nil
    playerNode = nil
  }
  
  // MARK: - DSP Helpers (Stubs)
  
  private func generateWaveform(type: String, frequency: Double, sampleCount: Int, sampleRate: Double) -> [Float] {
    // TODO: Implement waveform generation
    // Use vDSP for efficient computation
    return []
  }
  
  private func applyEnvelope(buffer: inout [Float], fadeInSamples: Int, fadeOutSamples: Int) {
    // TODO: Implement ADSR envelope
  }
}

File: modules/expo-audio-engine/android/ExpoAudioEngineModule.kt
/**
 * Android Native Audio Engine Module
 * 
 * Uses AudioTrack for low-latency audio synthesis.
 * Implements real-time tone generation, sweeps, and multi-tone playback.
 * 
 * IMPLEMENTATION GUIDE:
 * 
 * 1. Setup AudioTrack with optimal buffer size
 * 2. Generate PCM samples for different waveforms
 * 3. Use AudioTrack.write() for streaming playback
 * 4. Handle audio focus and routing
 * 5. Monitor audio device changes
 * 
 * Key Classes:
 * - AudioTrack: Main playback class
 * - AudioFormat: Sample rate, encoding, channels
 * - AudioAttributes: Audio usage and content type
 * - AudioManager: System audio management
 * 
 * Performance Tips:
 * - Use MODE_STREAM for continuous playback
 * - Buffer size: AudioTrack.getMinBufferSize() * 2
 * - Use FloatArray for samples (ENCODING_PCM_FLOAT)
 * - Pre-generate waveforms when possible
 * - Use Kotlin coroutines for async operations
 */

package expo.modules.audioengine

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlin.math.*

class ExpoAudioEngineModule : Module() {
  private var audioTrack: AudioTrack? = null
  private var isPlaying = false
  private var sampleRate = 48000
  private var bufferSize = 0
  private val scope = CoroutineScope(Dispatchers.IO)
  
  override fun definition() = ModuleDefinition {
    Name("ExpoAudioEngineModule")
    
    // Initialize
    AsyncFunction("initialize") { config: Map<String, Any> ->
      initializeAudioEngine(config)
    }
    
    // Play tone
    AsyncFunction("playTone") { params: Map<String, Any> ->
      playTone(params)
    }
    
    // Play sweep
    AsyncFunction("playSweep") { params: Map<String, Any> ->
      playSweep(params)
    }
    
    // Play multi-tone
    AsyncFunction("playMultiTone") { params: Map<String, Any> ->
      playMultiTone(params)
    }
    
    // Stop
    AsyncFunction("stop") {
      stopPlayback()
    }
    
    // Dispose
    AsyncFunction("dispose") {
      disposeAudioEngine()
    }
  }
  
  // MARK: - Audio Engine Setup
  
  private fun initializeAudioEngine(config: Map<String, Any>) {
    // Extract config
    sampleRate = (config["sampleRate"] as? Double)?.toInt() ?: 48000
    val requestedBufferSize = (config["bufferSize"] as? Double)?.toInt() ?: 512
    
    // Calculate buffer size
    val minBufferSize = AudioTrack.getMinBufferSize(
      sampleRate,
      AudioFormat.CHANNEL_OUT_MONO,
      AudioFormat.ENCODING_PCM_FLOAT
    )
    bufferSize = maxOf(minBufferSize, requestedBufferSize * 4) // Float = 4 bytes
    
    // Create AudioAttributes
    val audioAttributes = AudioAttributes.Builder()
      .setUsage(AudioAttributes.USAGE_MEDIA)
      .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
      .build()
    
    // Create AudioFormat
    val audioFormat = AudioFormat.Builder()
      .setSampleRate(sampleRate)
      .setEncoding(AudioFormat.ENCODING_PCM_FLOAT)
      .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
      .build()
    
    // Create AudioTrack
    audioTrack = AudioTrack.Builder()
      .setAudioAttributes(audioAttributes)
      .setAudioFormat(audioFormat)
      .setBufferSizeInBytes(bufferSize)
      .setTransferMode(AudioTrack.MODE_STREAM)
      .build()
  }
  
  // MARK: - Playback Methods (Stubs)
  
  private fun playTone(params: Map<String, Any>) {
    // TODO: Implement tone generation
    // 1. Extract parameters
    // 2. Generate waveform samples
    // 3. Apply envelope
    // 4. Write to AudioTrack
    // 5. Play
    
    scope.launch {
      audioTrack?.let { track ->
        if (track.state == AudioTrack.STATE_INITIALIZED) {
          track.play()
          isPlaying = true
          
          // Generate and write samples
          // val samples = generateTone(...)
          // track.write(samples, 0, samples.size, AudioTrack.WRITE_BLOCKING)
          
          track.stop()
          isPlaying = false
        }
      }
    }
  }
  
  private fun playSweep(params: Map<String, Any>) {
    // TODO: Implement frequency sweep
    // Similar to playTone but with changing frequency
  }
  
  private fun playMultiTone(params: Map<String, Any>) {
    // TODO: Implement multi-tone synthesis
    // Generate multiple waveforms and mix
  }
  
  private fun stopPlayback() {
    audioTrack?.stop()
    isPlaying = false
  }
  
  private fun disposeAudioEngine() {
    stopPlayback()
    audioTrack?.release()
    audioTrack = null
  }
  
  // MARK: - DSP Helpers (Stubs)
  
  private fun generateWaveform(
    type: String,
    frequency: Double,
    sampleCount: Int,
    sampleRate: Int
  ): FloatArray {
    // TODO: Implement waveform generation
    val samples = FloatArray(sampleCount)
    
    when (type) {
      "sine" -> {
        for (i in 0 until sampleCount) {
          val phase = 2.0 * PI * frequency * i / sampleRate
          samples[i] = sin(phase).toFloat()
        }
      }
      "square" -> {
        for (i in 0 until sampleCount) {
          val phase = 2.0 * PI * frequency * i / sampleRate
          samples[i] = if (sin(phase) > 0) 1.0f else -1.0f
        }
      }
      // Add more waveforms...
    }
    
    return samples
  }
  
  private fun applyEnvelope(
    samples: FloatArray,
    fadeInSamples: Int,
    fadeOutSamples: Int
  ) {
    // TODO: Implement envelope
    // Apply fade in
    for (i in 0 until minOf(fadeInSamples, samples.size)) {
      samples[i] *= (i.toFloat() / fadeInSamples)
    }
    
    // Apply fade out
    val start = maxOf(0, samples.size - fadeOutSamples)
    for (i in start until samples.size) {
      val factor = (samples.size - i).toFloat() / fadeOutSamples
      samples[i] *= factor
    }
  }
}

